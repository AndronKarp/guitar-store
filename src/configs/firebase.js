import firebase from "firebase";

const config = {
  apiKey: "AIzaSyDrF2CjklO0SlyV6Kv7pg7p67a4sLV3qbw",
  authDomain: "project-a0c8e.firebaseapp.com",
  databaseURL: "https://project-a0c8e.firebaseio.com",
  projectId: "project-a0c8e",
  storageBucket: "project-a0c8e.appspot.com",
  messagingSenderId: "686927330962",
  appId: "1:686927330962:web:6892625786538e6588304b",
  measurementId: "G-P6Q64LKV8Q"
};

firebase.initializeApp(config);

const database = firebase.database();
const guitarsRef = database.ref("guitars");
const auth = firebase.auth();
const usersRef = database.ref("users");
auth.getCurrentUser = () => {
  return new Promise((resolve, reject) => {
    const unsubscribe = firebase.auth().onAuthStateChanged(user => {
      unsubscribe();
      resolve(user);
    }, reject);
  });
};

export { guitarsRef, auth, usersRef };
