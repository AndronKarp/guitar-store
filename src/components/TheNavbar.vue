<template>
  <b-navbar sticky type="dark" variant="info">
    <b-navbar-brand class="guitar-store-page-link" to="/"
      >Guitar Store</b-navbar-brand
    >

    <b-navbar-toggle target="nav-collapse"></b-navbar-toggle>

    <b-collapse id="nav-collapse" is-nav>
      <b-navbar-nav class="ml-auto">
        <template v-if="currentUser">
          <b-nav-item class="shopping-cart-page-link" to="/cart">
            <b-icon-cart4></b-icon-cart4>
          </b-nav-item>
          <b-dropdown :text="currentUser.displayName" variant="info">
            <b-dropdown-item @click="signOut">Sign Out</b-dropdown-item>
          </b-dropdown>
        </template>

        <template v-else>
          <b-button
            class="sign-up-button mr-2"
            to="/join"
            squared
            variant="info"
            >Sign Up</b-button
          >
          <b-button
            class="sign-in-button"
            to="/login"
            squared
            variant="outline-light"
            >Sign In</b-button
          >
        </template>
      </b-navbar-nav>
    </b-collapse>
  </b-navbar>
</template>

<script>
import { mapGetters } from "vuex";
import { auth } from "../configs/firebase";
import moveToPage from "../mixins/move-to-page";

export default {
  computed: {
    ...mapGetters(["currentUser"])
  },
  methods: {
    async signOut() {
      await auth.signOut();
      this.moveTo("PageSignIn");
      this.$store.dispatch("clearCart");
    }
  },
  mixins: [moveToPage]
};
</script>

<style></style>
