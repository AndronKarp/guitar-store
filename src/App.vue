<template>
  <div id="app">
    <TheNavbar />
    <router-view></router-view>
  </div>
</template>

<script>
import TheNavbar from "./components/TheNavbar";
import { guitarsRef } from "./configs/firebase";

export default {
  created() {
    this.setFirebaseEvents();
  },
  methods: {
    setFirebaseEvents() {
      guitarsRef.on("child_added", snapshot => {
        this.$store.dispatch("addGuitar", snapshot.val());
      });
    }
  },
  components: {
    TheNavbar
  }
};
</script>

<style lang="scss">
#app {
  min-height: 100vh;
  height: 100%;
}
</style>
