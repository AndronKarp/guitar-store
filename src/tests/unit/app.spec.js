import { shallowMount, createLocalVue } from "@vue/test-utils";
import App from "@/App";
import Vuex from "vuex";
import { BootstrapVue } from "bootstrap-vue";
import TheNavbar from "@/components/TheNavbar";
import { auth } from "@/configs/firebase";

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(BootstrapVue);

describe("App.vue", () => {
  let actions;
  let store;
  let wrapper;
  let setAuthObserver;
  let dispatch;

  beforeEach(() => {
    auth.onAuthStateChanged = jest.fn(callback => callback({ uid: "uid " }));
    actions = {
      fetchGuitars: jest.fn(),
      setGuitarsRefChildChangedObserver: jest.fn(),
      updateCurrentUser: jest.fn(),
      updateCartId: jest.fn(),
      fetchCart: jest.fn()
    };
    store = new Vuex.Store({
      actions
    });
    setAuthObserver = jest.spyOn(App.methods, "setAuthObserver");
    dispatch = jest.spyOn(store, "dispatch");
    wrapper = shallowMount(App, {
      localVue,
      store,
      stubs: ["router-view"]
    });
  });

  test("renders app", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("accepts TheNavbar child component", () => {
    const theNavbar = wrapper.findComponent(TheNavbar);
    expect(theNavbar.exists()).toBe(true);
  });

  describe("methods", () => {
    describe("setAuthObserver sets auth observer which", () => {
      test("invokes updateCurrentUser action with user object", () => {
        wrapper.vm.setAuthObserver();
        expect(dispatch).toHaveBeenCalledWith(
          "updateCurrentUser",
          auth.onAuthStateChanged(user => user)
        );
      });

      test("invokes loadUserCart method if user is authorized", () => {
        const loadUserCart = jest.spyOn(wrapper.vm, "loadUserCart");
        wrapper.vm.setAuthObserver();
        expect(loadUserCart).toHaveBeenCalledWith(
          expect.objectContaining({
            cartId: auth.onAuthStateChanged(user => user.uid)
          })
        );
      });

      test("doesn't invoke loadUserCart method if user is unauthorized", () => {
        auth.onAuthStateChanged = jest.fn(callback => callback(null));
        const loadUserCart = jest.spyOn(wrapper.vm, "loadUserCart");
        wrapper.vm.setAuthObserver();
        expect(loadUserCart).not.toHaveBeenCalled();
      });
    });

    describe("loadUserCart", () => {
      test("invokes updateCartId action", () => {
        const obj = { cartId: "id" };
        wrapper.vm.loadUserCart(obj);
        expect(dispatch).toHaveBeenCalledWith("updateCartId", obj.cartId);
      });

      test("invokes fetchCart action", () => {
        wrapper.vm.loadUserCart({ cartId: "id" });
        expect(dispatch).toHaveBeenCalledWith("fetchCart");
      });
    });

    describe("lifecycle methods", () => {
      describe("created", () => {
        test("invokes setAuthObserver method", () => {
          expect(setAuthObserver).toHaveBeenCalled();
        });

        test("invokes fetchGuitars action", () => {
          expect(dispatch).toHaveBeenCalledWith("fetchGuitars");
        });

        test("invokes setGuitarsRefChildChangedObserver action", () => {
          expect(dispatch).toHaveBeenCalledWith(
            "setGuitarsRefChildChangedObserver"
          );
        });
      });
    });
  });
});
