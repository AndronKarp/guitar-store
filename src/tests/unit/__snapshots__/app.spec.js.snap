// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue renders app 1`] = `
<div id="app">
  <thenavbar-stub></thenavbar-stub>
  <b-container-stub tag="div" fluid="true" class="d-flex flex-column flex-fill mt-3">
    <router-view-stub></router-view-stub>
  </b-container-stub>
</div>
`;
